### PR TITLE
dhall: add --output options to dhall

### PR DESCRIPTION
**Pull request by Rizary**
As requested in #1386.

### CODE REVIEW

File: dhall/src/Dhall/Util.hs [hunk 1]
```diff
@@ -120,6 +121,9 @@ data Censor = NoCensor | Censor
 -- | Path to input
 data Input = StandardInput | File FilePath
 
+-- | Path to output
+data Output = StandardOutput | Output FilePath
```
Comment (sjakobi) [inline dhall/src/Dhall/Util.hs]:
```suggestion
data Output = StandardOutput | OutputFile FilePath
```

File: dhall/src/Dhall/Main.hs [hunk 2]
```diff
@@ -391,7 +405,7 @@ command (Options {..}) = do
 
     let renderDoc :: Handle -> Doc Ann -> IO ()
         renderDoc h doc = do
-            let stream = Pretty.layoutSmart layoutOpts doc
+            let stream = Pretty.layoutSmart layoutOpts doc
```
Comment (sjakobi) [inline dhall/src/Dhall/Main.hs]:
```suggestion
            let stream = Pretty.layoutSmart layoutOpts doc
```

File: dhall/src/Dhall/Main.hs [hunk 3]
```diff
@@ -440,7 +454,9 @@ command (Options {..}) = do
                         then Annot alphaNormalizedExpression inferredType
                         else alphaNormalizedExpression
 
-            render System.IO.stdout annotatedExpression
+            case output of
+                StandardOutput -> render System.IO.stdout annotatedExpression
+                Output file_ -> System.IO.withFile file_ System.IO.WriteMode $ (flip render) annotatedExpression
```
Comment (sjakobi) [inline dhall/src/Dhall/Main.hs]: Can you move the code after `->` to the next line and write it without `flip`? The easier to read the better!

File: dhall/src/Dhall/Main.hs [hunk 4]
```diff
@@ -225,6 +227,18 @@ parseMode =
...
                 <>  Options.Applicative.metavar "FILE"
                 )
+    
+    parseOutput = fmap f (optional p)
+      where
+        f Nothing = StandardOutput
+        f (Just file) = Output file
+
+        p = Options.Applicative.strOption
+                (   Options.Applicative.short 'o'
```
Comment (sjakobi) [inline dhall/src/Dhall/Main.hs]: So far `dhall` has largely refrained from using `short` options – I'm not sure why though. Probably we want the CLI commands to be as readable as possible too…

Thoughts, @Gabriel439?

File: dhall/src/Dhall/Main.hs (hunk 3 above)
Comment (Rizary) [reply dhall/src/Dhall/Main.hs]: Regarding the `flip`, sure. But it will also caused several changes to other lines. Is it okay?

Comment (sjakobi) [reply dhall/src/Dhall/Main.hs]: I'm not quite sure what you mean, but please give it a try.

File: dhall/src/Dhall/Main.hs (hunk 4 above)
Comment (Rizary) [reply dhall/src/Dhall/Main.hs]: Okay, I'll just remove it.

File: dhall/src/Dhall/Main.hs [hunk 5]
```diff
@@ -440,7 +453,10 @@ command (Options {..}) = do
                         then Annot alphaNormalizedExpression inferredType
                         else alphaNormalizedExpression
 
-            render System.IO.stdout annotatedExpression
+            case output of
+                StandardOutput -> render annotatedExpression System.IO.stdout
+                OutputFile file_ -> 
+                    System.IO.withFile file_ System.IO.WriteMode $ render annotatedExpression
```
Comment (sjakobi) [inline dhall/src/Dhall/Main.hs]: Ah, sorry for the misunderstanding. I thought of something like

```suggestion
                    System.IO.withFile file_ System.IO.WriteMode $ \h -> render h annotatedExpression
```

File: dhall/src/Dhall/Main.hs (hunk 4 above)
Comment (Gabriel439) [reply dhall/src/Dhall/Main.hs]: Yeah, the use of only long options is intentional.

In my experience, short options are an impediment to learning to beginners (especially if they learn from tutorial material that uses short options or by observing other people using short options).  Short options also tend to lead to people cargo-culting the options blindly without thinking about what they do.  This comment accurately summarizes what happens when short options become idiomatic:

https://xkcd.com/1168/

**Pull request merged**